File: js/zombie/target.js
goog.provide('energy.zombie.Target');


goog.require('goog.dom');
goog.require('goog.dom.classes');
goog.require('goog.style');
goog.require('goog.events');
goog.require('goog.userAgent');
goog.require('goog.userAgent.product');

goog.require('energy.sound');

/**
 * @constructor
 */
energy.zombie.Target = function(game) {
  this.game = game;
};

/**
 * @type {Element}
 */
energy.zombie.Target.prototype.targetElement = null;

/**
 * @type {energy.zombie.Game}
 */
energy.zombie.Target.prototype.game = null;

/**
 * @type {number}
 */
energy.zombie.Target.angleIncrese = 15;

/**
 * @type {number}
 */
energy.zombie.Target.prototype.turns = 0;


/**
 * @type {number}
 */
energy.zombie.Target.prototype.stopTime = 0;


/**
 * @type {number}
 */
energy.zombie.Target.prototype.currentTurn = 0;

/**
 * @type {boolean}
 */
energy.zombie.Target.prototype.front = true;


/**
 * @type {boolean}
 */
energy.zombie.Target.prototype.rotating = false;

/**
 * @type {number}
 */
energy.zombie.Target.prototype.angle = 0;

energy.zombie.Target.prototype.update = function() {
  if (this.rotating) {
    this.continueRotating();
  }
};

energy.zombie.Target.prototype.continueRotating = function() {
  if (this.angle >= 360) {
    this.angle %= 360;
  }
  var className = 'target';

  if (this.angle % 180 == 0) {
    this.currentTurn++;
    if (this.currentTurn > this.turns) {
      this.stopTime = new Date().getTime();
      this.rotating = false;
    }
  }

  if (this.currentTurn <= this.turns) {
    if (this.angle < 90 && this.angle + energy.zombie.Target.angleIncrese >= 90) {
      goog.dom.classes.remove(this.targetElement, 'front');
      goog.dom.classes.remove(this.targetElement, 'target-hit');
      goog.dom.classes.add(this.targetElement, 'target');

      var number = Math.ceil(Math.random() * 3);
      goog.dom.classes.add(this.targetElement, className + '-' + number);
      this.front = false;
    }
    if (this.angle < 270 && this.angle + energy.zombie.Target.angleIncrese >= 270) {
      goog.dom.classes.add(this.targetElement, 'front target');
      //goog.dom.classes.remove(this.targetElement, 'target');
      for (var i = 1; i <=3; i++) {
        goog.dom.classes.remove(this.targetElement, 'target-' + i);
      }
      goog.dom.classes.remove(this.targetElement, 'target-hit');

      this.front = true;
    }

    if (goog.userAgent.WEBKIT) {
      goog.style.setStyle(this.targetElement, '-webkit-transform', 'rotateY(' + this.angle + 'deg)');
    } else if (goog.userAgent.GECKO) {
      goog.style.setStyle(this.targetElement, '-moz-transform', 'scale(' + Math.cos(this.angle * Math.PI / 180) + ', 1)');
    } else if (goog.userAgent.IE) {
      goog.style.setStyle(this.targetElement, '-ms-transform', 'scale(' + Math.cos(this.angle * Math.PI / 180) + ', 1)');
    } else {
      goog.style.setStyle(this.targetElement, '-o-transform', 'scale(' + Math.cos(this.angle * Math.PI / 180) + ', 1)');
    }

    this.angle += energy.zombie.Target.angleIncrese;
  }
};

energy.zombie.Target.prototype.draw = function(container) {
  this.targetElement = goog.dom.createElement('div');
  goog.dom.classes.add(this.targetElement, 'target');
  goog.dom.classes.add(this.targetElement, 'front');
  goog.dom.appendChild(container, this.targetElement);

  if (goog.userAgent.product.IPAD || goog.userAgent.product.ANDROID) {
    goog.events.listen(container, goog.events.EventType.TOUCHSTART, this.listener, true, this);
  } else {
    goog.events.listen(container, goog.events.EventType.MOUSEDOWN, this.listener, true, this);
  }
};

/**
 * @return {boolean}
 */
energy.zombie.Target.prototype.isRotating = function() {
  return this.rotating;
};

energy.zombie.Target.prototype.rotate = function(turns) {
  this.rotating = true;
  this.turns = turns;
  this.currentTurn = 0;
};


energy.zombie.Target.prototype.listener = function(e) {
  if (!e.isButton(goog.events.BrowserEvent.MouseButton.LEFT)) return;
  var score = this.game.score;
  var className = null;
  if (this.front) {
    className = 'shrink';
    score = Math.max(0, this.game.score - 5);
    energy.sound.play("fail", 'ogg');
  } else {
    className = 'grow';
    this.rotate(1);
    score = this.game.score + 10;
    goog.dom.classes.add(this.targetElement, 'target-hit');
    energy.sound.play('hit', 'ogg');
  }

  if (score != this.score) {
    this.game.score = score;
    var scoreElement = goog.dom.getElement('score');
    goog.dom.setTextContent(scoreElement, score);
    scoreElement.className = '';
    goog.dom.classes.add(scoreElement, className);
  }
  e.preventDefault();
  return false;
};

File: src/main/resources/static/scripts/target.dart
import 'dart:html';
import 'game.dart';
import 'dart:math' as math;
import 'dart:async';

import 'sound.dart';
import 'stats.dart';

class Target {

  static const int TARGET_DELAY = 2000;
  static const int ANGLE_DELTA = 15;

  Game game;
  DivElement element;
  DivElement remainingElement;
  StreamSubscription listener;
  StreamSubscription touchListener;

  bool rotating = false;
  bool front = true;
  int angle = 0;
  int turns = 0;
  int currentTurn = 0;
  int started = 0;
  Stopwatch stopwatch = new Stopwatch();

  Target(Game game) {
    this.game = game;
  }

  DivElement draw() {
    element = new DivElement();
    element.classes.add('target');
    element.classes.add('front');

    remainingElement = new DivElement();
    remainingElement.classes.add('remaining');

    element.children.add(remainingElement);

    return element;
  }

  void update() {
    if (rotating) {
      continueRotation();
    } else {
      if (!front) {
        if (!stopwatch.isRunning) {stopwatch.reset();
          stopwatch.start();
          remainingElement.style.display = 'block';
        } else if (stopwatch.elapsedMilliseconds > TARGET_DELAY) {
          remainingElement.style.display;
          hideRemaining();
          stopwatch.stop();
          if (game.health.attack()) {
            Sound.play('attack');
          }
          rotate(1);
        } else {
          String scaleX = (1 - stopwatch.elapsedMilliseconds / TARGET_DELAY).toStringAsFixed(2);
          remainingElement.style.transform  = 'scale(${scaleX}, 1)';
        }
      }
    }
  }

  void hideRemaining() {
    remainingElement.style.transform = '';
    remainingElement.style.display = 'none';
  }

  void listen() {
    var handler = (e) {
      if (e is MouseEvent && (e as MouseEvent).button != 0) {
        e.preventDefault();
        return false;
      }
      int score = game.score;
      String className = null;
      if (front) {
        className = 'shrink';
        score = math.max(0, game.score - 5);
        Sound.play('fail');
        Stats.miss++;
        game.multiplier.miss();
      } else {
        className = 'grow';
        rotate(1);
        score = game.score + 10 * game.multiplier.factor;
        element.classes.add('target-hit');
        Stats.hit++;
        game.multiplier.hit();
        Sound.play('hit');
        hideRemaining();
      }
      stopwatch.stop();

      if (score != game.score) {
        game.score = score;
      }
      e.preventDefault();
      return false;
    };

    touchListener = element.onTouchStart.listen(handler);
    listener = element.onMouseDown.listen(handler);
  }


  void stopListening() {
    touchListener.cancel();
    listener.cancel();
  }

  void rotate(int _turns) {
    stopwatch.reset();
    rotating = true;
    turns = _turns;
    currentTurn = 0;
  }

  void continueRotation() {
    if (angle >= 360) {
      angle %= 360;
    }

    if (angle % 180 == 0) {
      currentTurn++;
      if (currentTurn > turns) {
        rotating = false;
      }
    }

    if (currentTurn <= turns) {
      if (angle < 90 && angle + ANGLE_DELTA >= 90) {
        element.classes.remove('front');
        element.classes.remove('target-hit');
        element.classes.add('target');
        element.classes.add('target-${(new math.Random().nextDouble() * 3).ceil()}');
        this.front = false;
      }
      if (angle < 270 && angle + ANGLE_DELTA >= 270) {
        element.classes.clear();
        element.classes.add('front');
        element.classes.add('target');
        this.front = true;
      }
      angle += ANGLE_DELTA;
      element.style.transform = 'rotateY(${angle}deg)';
    }
  }

  void flipRandomly(bool even) {
    int flips = (new math.Random().nextDouble() * 3).round();
    this.rotate(flips % 2 == (even ? 1 : 0) ? flips : flips + 1);
  }
}